File: src/save.py
import sys
import os
import shutil
import time
import argparse
import zipfile
import platform
import concurrent.futures
import pdb

debugMode = False

def log(message):
    if debugMode:
        print(message)

def timeFormat():
    currentTime = time.time()
    formattedTime = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(currentTime))
    return formattedTime

def generatePaths(steamLibrary, localLibrary, gameName, savePath, backupPath):
    if not "~" in savePath:
        src = os.path.normpath(f"{steamLibrary}/{savePath}")
    else:
        src = os.path.normpath(f"{savePath}")
        src = os.path.expanduser(src)

    targ = os.path.normpath(f"{localLibrary}{gameName}/{timeFormat()}/{backupPath}")
    return src, targ

def performCopy(src, targ, gameName, dryRun=False):
    if dryRun == False:
        log(f"Saving files for '{gameName}'")
        shutil.copytree(src, targ)
        zipPath = targ
        shutil.make_archive(zipPath, 'zip', targ)
        log(f"Creating zip archive of '{gameName}'")
        shutil.rmtree(targ)

def saveGame(steamLibrary, localLibrary, maxBackups, option, path):
    try:
        log(path)
        gameName, savePath, backupPath = path.split("|")
    except ValueError:
        log("Value Error")
        quit("Value Error")

    if option == None:
        quit("Please select save or restore")

    src, targ = generatePaths(steamLibrary, localLibrary, gameName, savePath, backupPath)
    if not os.path.exists(f"{src}"):
        log(f"You do not appear to have the game '{gameName}'")
        return

    if not os.path.exists(f"{localLibrary}{gameName}"): # If save path does not exist
        # This is here because you get things like <folder>/../../<folder> and this errors for some reason
        performCopy(src, targ, gameName)
        log(f"Saved data for {gameName}")
    else:
        if option.lower() == "save":
            zipFiles = [f for f in os.listdir(f"{localLibrary}{gameName}") if f.endswith(".zip")]
            zipFiles = sorted(zipFiles)
            log(f"You have {len(zipFiles)} backups for game: '{gameName}'")
        
            while len(zipFiles) > maxBackups: # While loop because if you lower the amount of backups that you want saved you want all the old ones deleted
                log(f"More than {maxBackups} backups for game '{gameName}'. Removing the oldest: {zipFiles[-1]}")
                oldestBackup = os.path.join(f"{localLibrary}{gameName}", zipFiles[-1])
                os.remove(oldestBackup)
            performCopy(src, targ, gameName)
            log(f"Saved data for {gameName}")
        if option.lower() == "restore":
            quit("Not yet implimented") # TODO

    return

def saveGames(steamLibrary, localLibrary, maxBackups, option):
    log(platform.system())
    try:
        readlines = open(f"../SavePathDataset-{platform.system()}.txt", 'r').read().splitlines()
    except:
        quit(f"You do not have any datasets for platform: {platform.system()}")
    savePaths = []
    for path in readlines:
        if "**" in path and not path.startswith("**"):
            path = path.split("**") # This is for end of the line comments
            path = path[0]
        if not path.startswith("**") and not path == "": # If not a comment and not empty
            savePaths.append(path)
    numberOfGamesToSave = len(savePaths)
    log(f"Searching {numberOfGamesToSave} game save data locations.")
    if not debugMode:
        with concurrent.futures.ThreadPoolExecutor() as executor:
            futures = [executor.submit(saveGame, steamLibrary, localLibrary, maxBackups, option, path) for path in savePaths]
    else:
        for path in savePaths:
            saveGame(steamLibrary, localLibrary, maxBackups, option, path)

def pushFiles():
    pass

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Used to save and restore steam game save data')
    parser.add_argument('steamLibraryPath', nargs='?', help='The path to your steam library')
    parser.add_argument('-s', '--save', action='store_true', help='Save Steam game data')
    parser.add_argument('-r', '--restore', action='store_true', help='Restore Steam game data')
    parser.add_argument('-l', '--localLibrary', nargs='?', help='Location to store or restore Steam data from or to')
    parser.add_argument('-b', '--backups', type=int, help='The number of Steam game backups to store during saving')
    parser.add_argument('-d', '--debug', action='store_true', help='Used by developers to debug the program')
    args = parser.parse_args()
    # example steamLibrary path /media/<user>/<drive>/steamLibrary/
    # add argParser with required steamLibrary path but other optional
    steamLibrary = None
    print(args)

    if args.debug:
        debugMode = True

    if args.backups == None:
        args.backups = 2 # default
    if args.localLibrary == None:
        args.localLibrary = "../SteamSaveLocal/"

    if not os.path.exists(args.localLibrary):
        os.mkdir(args.localLibrary)

    option = None
    if args.save:
        option = "save"

    if args.restore:
        option = "restore"

    saveGames(args.steamLibraryPath, args.localLibrary, args.backups, option)

File: src/save.go
package main

import (
    "archive/zip"
    "fmt"
    "io"
    "os"
    "path/filepath"
    "strings"
    "time"
)

func log(message string) {
    fmt.Println(message)
}

func timeFormat() string {
    currentTime := time.Now()
    formattedTime := currentTime.Format("2006-01-02 15:04:05")
    return formattedTime
}

func copyDir(src string, dst string) error {
    srcFi, err := os.Stat(src)
    if err != nil {
        return err
    }

    err = os.MkdirAll(dst, srcFi.Mode())
    if err != nil {
        return err
    }

    entries, err := os.ReadDir(src)
    if err != nil {
        return err
    }

    for _, entry := range entries {
        srcPath := filepath.Join(src, entry.Name())
        dstPath := filepath.Join(dst, entry.Name())

        if entry.IsDir() {
            err = copyDir(srcPath, dstPath)
            if err != nil {
                return err
            }
        } else {
            err = copyFile(srcPath, dstPath)
            if err != nil {
                return err
            }
        }
    }

    return nil
}

func copyFile(src string, dst string) error {
    in, err := os.Open(src)
    if err != nil {
        return err
    }
    defer in.Close()

    out, err := os.Create(dst)
    if err != nil {
        return err
    }
    defer out.Close()

    _, err = io.Copy(out, in)
    if err != nil {
        return err
    }

    return out.Close()
}

func createZipArchive(src string, dst string) error {
    zipFile, err := os.Create(fmt.Sprintf("%s.createZipArchive", dst))
    if err != nil {
        return err
    }
    defer zipFile.Close()

    archive := zip.NewWriter(zipFile)
    defer archive.Close()

    err = filepath.Walk(src, func(path string, info os.FileInfo, err error) error {
        if err != nil {
            return err
        }

        relPath, err := filepath.Rel(src, path)
        if err != nil {
            return err
        }

        if info.IsDir() {
            return nil
        }

        header, err := zip.FileInfoHeader(info)
        if err != nil {
            return err
        }

        header.Name = relPath

        writer, err := archive.CreateHeader(header)
        if err != nil {
            return err
        }

        file, err := os.Open(path)
        if err != nil {
            return err
        }
        defer file.Close()

        _, err = io.Copy(writer, file)
        return err
    })
    return err
}

func saveGame(SteamLibrary string, LocalLibrary string, maxBackups int, option string, path string) {
    parts := strings.Split(path, "|")
    if len(parts) != 3 {
        return
    }

    gameName, savePath, backupPath := parts[0], parts[1], parts[2]
    src, targ := generatePaths(SteamLibrary, gameName, savePath, backupPath, LocalLibrary)

    if _, err := os.Stat(src); os.IsNotExist(err) {
        return
    }

    if _, err := os.Stat(fmt.Sprintf("%s%s", LocalLibrary, gameName)); os.IsNotExist(err) {
        performCopy(src, targ, gameName, false)
        log(fmt.Sprintf("Saved data for %s", gameName))
    } else {
        if strings.ToLower(option) == "all" {
            entries, err := os.ReadDir(fmt.Sprintf("%s%s", LocalLibrary, gameName))
            if err != nil {
                log(fmt.Sprintf("Error reading directory: %v", err))
                return
            }
            zipFiles := []string{}
            for _, file := range entries {
                if strings.HasSuffix(file.Name(), ".createZipArchive") {
                    zipFiles = append(zipFiles, file.Name())
                }
            }
            // sort zipFiles
            if len(zipFiles) > maxBackups {
                log(fmt.Sprintf("You have %d backups for game: '%s'", len(zipFiles), gameName))
                for len(zipFiles) > maxBackups {
                    log(fmt.Sprintf("More than %d backups for game '%s'. Removing the oldest: %s", maxBackups, gameName, zipFiles[len(zipFiles)-1]))
                    oldestBackup := filepath.Join(fmt.Sprintf("%s%s", LocalLibrary, gameName), zipFiles[len(zipFiles)-1])
                    err := os.Remove(oldestBackup)
                    if err != nil {
                        log(fmt.Sprintf("Error removing file: %v", err))
                        return
                    }
                }
                performCopy(src, targ, gameName, false)
                log(fmt.Sprintf("Saved data for %s", gameName))
            }
        }
    }
}

func saveGames(SteamLibrary string, LocalLibrary string, maxBackups int, option string) {
    file, err := os.Open("../SavePathDatasetLinux.txt")
    if err != nil {
        log(fmt.Sprintf("Error opening file: %v", err))
        return
    }
    defer file.Close()

    // SavePaths := []string{}
}

func generatePaths(SteamLibrary string, gameName string, savePath string, backupPath string, LocalLibrary string) (string, string) {
    var src string
    if !strings.Contains(savePath, "~") {
        src = filepath.FromSlash(fmt.Sprintf("%s/%s", SteamLibrary, savePath))
    } else {
        src = filepath.FromSlash(savePath)
        src, _ = filepath.Abs(src)
    }

    targ := filepath.FromSlash(fmt.Sprintf("%s%s/%s/%s", LocalLibrary, gameName, timeFormat(), backupPath))
    return src, targ
}

func performCopy(src string, targ string, gameName string, dryRun bool) {
    if !dryRun {
        log(fmt.Sprintf("Saving files for '%s'", gameName))
        err := os.MkdirAll(targ, os.ModePerm)
        if err != nil {
            log(fmt.Sprintf("Error creating directory: %v", err))
            return
        }
        err = copyDir(src, targ)
        if err != nil {
            log(fmt.Sprintf("Error copying directory: %v", err))
            return
        }
        log(fmt.Sprintf("Creating zip archive of '%s'", gameName))
        zipPath := targ
        err = createZipArchive(zipPath, targ)
        if err != nil {
            log(fmt.Sprintf("Error zipping directory: %v", err))
            return
        }
        os.RemoveAll(targ)
    }
}

func main() {
    SteamLibrary := "/media/robot/steamgames/SteamLibrary"
    maxBackups := 2
    option := ""
    if len(os.Args) > 2 {
        option = os.Args[1]
        SteamLibrary = os.Args[2]
    }
    if SteamLibrary == "" {
        log("Please provide SteamLibrary path")
        return
    }

    SteamLibrary = filepath.FromSlash(fmt.Sprintf("%ssteamapps/", SteamLibrary))

    LocalLibrary := ""
    if len(os.Args) > 3 {
        LocalLibrary = os.Args[3]
    }
    if LocalLibrary == "" {
        LocalLibrary = "../SteamSaveLocal/"
    }

    if _, err := os.Stat(LocalLibrary); os.IsNotExist(err) {
        err := os.Mkdir(LocalLibrary, os.ModePerm)
        if err != nil {
            log(fmt.Sprintf("Error creating directory: %v", err))
            return
        }
    }

    saveGames(SteamLibrary, LocalLibrary, maxBackups, option)
}
